0016-pre-shared-key/0016-pre-shared-key.md

```diff
@@ -0,0 +1,260 @@
+
+PSK payments use an HMAC of the PSK details as a fulfillment, to prove that the contents are unmodified. The pre-shared secret key is the key used in the HMAC.
+
+The data attached to these headers is an encrypted binary blob. This encrypted blob contains the private PSK headers.
+
+Despite the fact that connectors can read the public headers' data, connectors should not be required to do so, nor should they rely on PSK details being present on all transfers. This information is only intended for the initial sender and final receiver of a payment.
+
+The decryption key is derived from the pre-shared secret key, and the AES-256-CTR initialization vector is set to the value of the `Nonce` header (`fpwpAhlN588` in the above example).
+
+| Header | Value |
+|--------|-------|
+| `Key` | _(Required)_ The HMAC algorithm, which MUST be `hmac-sha-256` (case sensitive) |
+| `Nonce` | _(Required)_ A [base64url-encoded](https://en.wikipedia.org/wiki/Base64#URL_applications) nonce, used to generate ensure uniqueness of the PSK data and as an initialization vector (IV) for AES-256-CTR encryption. The nonce **MUST** be 16 bytes, and **MUST** be generated with strong randomness. |
+| `Encryption` | _(Required)_ MUST be set to `aes-256-ctr` or `none`. If it is set to `aes-256-ctr`, then private headers and application data will be AES-256-CTR encrypted. If it is set to `none`, then private headers and application data will be in cleartext. This cleartext will still be appended to the public headers after an empty line. |
+| ... | _(Optional)_ Additional headers. These can be read by any connectors. |
+
+### Private Headers
+
+The private headers and their data are formatted as below:
+
+```http
+Expires-At: 2017-03-13T16:58:06.901Z
+Header-A: some_stuff
+Header-B: some_other_stuff
+Header-C: some_more_stuff
+Date: 2017-03-13T16:57:56.901Z
+Content-Type: application/json
+
+{"foo":"bar"}
+```
+
+Only the payment's sender and receiver can decrypt and read the private headers and their data. The private headers contains an arbitrary message body preceded by the following headers:
+
+| Header | Value |
+|--------|-------|
+| `Expires-At` | _(Optional)_ An [ISO8601 UTC timestamp](https://en.wikipedia.org/wiki/ISO_8601), after which this PSK payment has expired. |
+| `Content-Type` | _(Optional)_ The [media type](http://www.iana.org/assignments/media-types/media-types.xhtml) of the data in the message body |
+| ... | _(Optional)_ Additional arbitrary headers. Where appropriate, it is RECOMMENDED to match the headers defined for HTTP messages ([RFC7230](https://tools.ietf.org/html/rfc7230#section-3.2)) |
+
+# Pseudocode
+
+The following pseudocode outlines the generation procedure for the necessary values involved in PSK. It follows only the case where encryption is enabled, for simplicity.
+
+### Receiver: Setup of the Protocol
+
```

Comment:
Include a note that this first code block is optional and not part of the protocol. The receiver may use any method they choose for generating and regenerating or fetching the `shared_secret`. The advantage of this method is that the receiver doesn't need to store a load of secrets, just one